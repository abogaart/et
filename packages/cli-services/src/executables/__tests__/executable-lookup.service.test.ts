import { ExecutableLookupService, ExecutableLookup, Executable } from '..';

function mockExecutable(name: string, path: string, version: string): any {
  return {
    getName: jest.fn().mockReturnValue(name),
    getPath: jest.fn().mockReturnValue(path),
    getVersion: jest.fn().mockReturnValue(version),
  };
}

class MockExecutableLookup extends ExecutableLookup {
  createExecutable(path: string): Executable {
    return mockExecutable(this.name, path, `${path}-version`);
  }
}

describe('Executable Lookup Service class', () => {
  let brdLookup: ExecutableLookup;

  beforeEach(() => {
    brdLookup = new MockExecutableLookup('brd');
  });

  describe('create', () => {
    it('creates an executable', () => {
      const service = new ExecutableLookupService();
      service.register('brd', brdLookup);

      const exec = service.create('brd', 'brd-path');
      expect(exec).not.toBeNull();
      expect(exec.getName()).toBe('brd');
      expect(exec.getPath()).toBe('brd-path');
      expect(exec.getVersion()).toBe('brd-path-version');
    });

    it('throws an error if executable lookup is not found', () => {
      const service = new ExecutableLookupService();
      try {
        service.create('brd', 'brd-path');
        fail('Should have thrown an error');
      } catch (e) {
        expect(e.message).toMatchSnapshot();
      }
    });
  });

  describe('lookup', () => {
    it('looks up an executable', async () => {
      const exec1 = mockExecutable('brd', 'brd-path', 'brd-version');
      const exec2 = mockExecutable('brd2', 'brd2-path', 'brd2-version');

      brdLookup.resolve = jest.fn().mockReturnValue([exec1, exec2]);
      const service = new ExecutableLookupService();
      service.register('brd', brdLookup);

      const execs = await service.lookup('brd');
      expect(execs).toEqual([exec1, exec2]);
    });

    it('returns zero executables if no executable lookup found', async () => {
      const service = new ExecutableLookupService();
      const execs = await service.lookup('brd');

      expect(execs).toHaveLength(0);
    });
  });
});
