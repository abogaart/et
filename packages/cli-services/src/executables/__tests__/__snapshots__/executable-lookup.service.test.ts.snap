// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Executable Lookup Service class create throws an error if executable lookup is not found 1`] = `"Failed to create executable brd with path brd-path. No executable lookup defined for brd."`;
